heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,267 @@
+  ignoreSuppressed: boolean;
+  eslintVersion?: string;
+}
+
+export interface ISarifRun {
+  tool: {
+    driver: {
+      name: string;
+      informationUri: string;
+      version?: string;
+      rules: IStaticAnalysisRules[];
+    };
+  };
+  artifacts?: ISarifFile[];
+  results?: ISarifRepresentation[];
+  invocations?: {
+    toolConfigurationNotifications: ISarifRepresentation[];
+    executionSuccessful: boolean;
+  }[];
+}
+
+export interface ISarifRepresentation {
+  level: string;
+  message: {
+    text: string;
+  };
+  locations: ISarifLocation[];
+  ruleId?: string;
+  descriptor?: {
+    id: string;
+  };
+  suppressions?: ISuppressedAnalysis[];
+}
+
+// Interface for the SARIF log structure
+export interface ISarifLog {
+  version: string;
+  $schema: string;
+  runs: ISarifRun[];
+}
+
+export interface IRegion {
+  startLine?: number;
+  startColumn?: number;
+  endLine?: number;
+  endColumn?: number;
+  snippet?: {
+    text: string;
+  };
+}
+
+export interface IStaticAnalysisRules {
+  id: string;
+  name?: string;
+  shortDescription?: {
+    text: string;
+  };
+  fullDescription?: {
+    text: string;
+  };
+  defaultConfiguration?: {
+    level: 'note' | 'warning' | 'error';
+  };
+  helpUri?: string;
+  properties?: {
+    category?: string;
+    precision?: 'very-high' | 'high' | 'medium' | 'low';
+    tags?: string[];
+    problem?: {
+      severity?: 'recommendation' | 'warning' | 'error';
+      securitySeverity?: number;
+    };
+  };
+}
+
+export interface ISarifFile {
+  location: {
+    uri: string;
+  };
+}
+
+export interface ISuppressedAnalysis {
+  kind: string;
+  justification: string;
+}
+
+export interface ISarifLocation {
+  physicalLocation: ISarifPhysicalLocation;
+}
+
+export interface ISarifPhysicalLocation {
+  artifactLocation: {
+    uri: string;
+    index: number;
+  };
+  region?: IRegion;
+}
+
+interface IMessage extends TEslint.Linter.LintMessage {
+  suppressions?: ISuppressedAnalysis[];
+}
+
+const internalErrorId: 'ESL0999' = 'ESL0999';
+
+/**
+ * Converts ESLint results into a SARIF (Static Analysis Results Interchange Format) log.
+ *
+ * This function takes in a list of ESLint lint results, processes them to extract
+ * relevant information such as errors, warnings, and suppressed messages, and
+ * outputs a SARIF log which conforms to the SARIF v2.1.0 specification.
+ *
+ * @param results - An array of lint results (`TEslint.ESLint.LintResult[]`) from ESLint that contains linting information,
+ *                  such as file paths, messages, and suppression details.
+ * @param options - An object (`ISerifFormatterOptions`) containing options for formatting:
+ *                  - `ignoreSuppressed`: Boolean flag to decide whether to ignore suppressed messages.
+ *                  - `eslintVersion`: Optional string to include the version of ESLint in the SARIF log.
+ * @returns The SARIF log (`ISarifLog`) containing information about the linting results in SARIF format.
```

Comment:
```suggestion
 * @param results - An array of lint results from ESLint that contains linting information,
 *                  such as file paths, messages, and suppression details.
 * @param options - An object containing options for formatting:
 *                  - `ignoreSuppressed`: Boolean flag to decide whether to ignore suppressed messages.
 *                  - `eslintVersion`: Optional string to include the version of ESLint in the SARIF log.
 * @returns The SARIF log containing information about the linting results in SARIF format.
```

The types are visible in the declaration files.